### PR TITLE
Share built images in CI and more sharding for functional tests

### PR DESCRIPTION
**Pull request by dnr**
**What changed?**
- In CI, build most of the tests and dependencies and reuse them in subsequent steps.
- Additionally pre-build dependencies separately on a weekly schedule (in buildkite configuration) and reuse them when possible.
- Add more test sharding for functional tests (split the main functional test suite into three).

**Why?**
Increase parallelism, reduce latency of getting test results, reduce granularity of retries

**How did you test it?**
lots of testing on buildkite

**Potential risks**

**Is hotfix candidate?**

### CODE REVIEW

**Review by mindaugasrukas — APPROVED** [review]

File: Makefile [hunk 1]
```diff
@@ -501,11 +499,13 @@ update-dashboards:
 gomodtidy:
 	@printf $(COLOR) "go mod tidy..."
 	@go mod tidy
+	@go list -deps $(FUNCTIONAL_TEST_ROOT) | sort -u | grep '^[a-z]\+\.[a-z.]\+/' | grep -v go.temporal.io > develop/buildkite/third_party_deps.txt
```
Comment (stephanos) [inline Makefile:502]: Why exclude the Go standard library packages; wouldn't we also benefit from pre-building them?

File: Makefile [hunk 2]
```diff
@@ -501,11 +499,13 @@ update-dashboards:
...
 
 update-dependencies:
 	@printf $(COLOR) "Update dependencies..."
 	@go get -u -t $(PINNED_DEPENDENCIES) ./...
 	@go mod tidy
+	@go list -deps $(FUNCTIONAL_TEST_ROOT) | sort -u | grep '^[a-z]\+\.[a-z.]\+/' | grep -v go.temporal.io > develop/buildkite/third_party_deps.txt
```
Comment (stephanos) [inline Makefile:508]: I'm not a Makefile expert, but would it make sense to move this to a separate target and call that via `$(MAKE) ...`? I'm just worried about having this non-trivial command in here twice.

File: develop/buildkite/prebuild-pipeline.yml
Comment (stephanos) [inline develop/buildkite/prebuild-pipeline.yml:1]: Would this have to be build outside of Buildkite by-hand or would there be a separate pipeline for this? (asking since it seems to be required for the pipeline to work, see `cache-from`; but I might be misunderstanding Buildkite)

File: develop/buildkite/prebuild-Dockerfile [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+ADD go.mod go.sum develop/buildkite/third_party_deps.txt /temporal/
+# Download deps (in several layers to parallelize push/pull).
+RUN go mod download github.com/aws/aws-sdk-go
+RUN go mod download google.golang.org/api
+RUN go mod download modernc.org/sqlite
+RUN go mod download
+
+# Build deps (in two layers).
+RUN head -100 third_party_deps.txt | xargs go build -race
+RUN tail +100 third_party_deps.txt | xargs go build -race
```
[the remark this answers is not in the document]
Comment (dnr) [reply develop/buildkite/prebuild-Dockerfile:15]: It does, it just builds the first 100 in one layer and the rest in another layer. It's to distribute layer size more evenly to get better docker push/pull parallelism. 100 was determined empirically and splits the size roughly in half (because the aws sdk packages are so big, and some of the earlier ones depend on later ones)

File: develop/buildkite/prebuild-pipeline.yml
Comment (dnr) [reply develop/buildkite/prebuild-pipeline.yml:1]: My plan is to change the buildkite config to do this:
```
steps:
  - command: "buildkite-agent pipeline upload develop/buildkite/${BK_PIPELINE_PREFIX}pipeline.yml"
    label: ":pipeline: Pipeline upload"
```
and then create a buildkite schedule that runs the build once a week with extra environment vars `BK_PIPELINE_PREFIX=prebuild-`. I didn't test it yet since it requires messing with the main config but I think it should it work

**Comment by dnr** [comment]
This should be pretty much ready for review. Before or after merging, we just need to increase max parallelism and add the scheduled build of dependencies.

File: Makefile [hunk 4]
```diff
@@ -9,7 +9,7 @@ bins: temporal-server temporal-cassandra-tool temporal-sql-tool tdbg
 all: update-tools clean proto bins check test
 
 # Used by Buildkite.
-ci-build: bins build-tests ci-update-tools shell-check copyright-check proto go-generate gomodtidy ensure-no-changes
+ci-build-misc: bins ci-update-tools shell-check copyright-check proto go-generate gomodtidy ensure-no-changes
```
Comment (yiminc) [inline Makefile:12]: why this change?

**Review by yiminc — APPROVED** [review]

File: Makefile (hunk 4 above)
Comment (dnr) [reply Makefile:12]: The "build-tests" part was basically just a quick check for compilation errors to avoid running the rest of the test suite (and then it threw away the results of course). Now there's a initial "build" step that builds the functional tests and all the parts of the server that they depend on (most of it), which basically serves the same purpose. The reason to build only functional tests and not unit tests in the build step is that the functional test compilation results will be reused by many steps, and steps that are on the critical path, while the unit test compilation results are only used by a single step, and that one is not on the critical path so having it do more work doesn't delay the overall run.
